### PR TITLE
implemented lazy loading of properties

### PR DESCRIPTION
**Pull request by danrot**
Fixes https://github.com/jackalope/jackalope/issues/303

Tests are running, but I will have a look at the tests of jackalope-doctrine-dbal and jackalope-jackrabbit as well. Will ping here again, when this is done.

### CODE REVIEW

File: src/Jackalope/Node.php [hunk 1]
```diff
@@ -191,15 +198,16 @@ private function parseData($rawData, $update, $keepChanges = false)
...
+                            if (isset($this->propertyValues[$key])) {
                                 // refresh existing binary, this will only happen in update
                                 // only update length
-                                if (! ($keepChanges && $this->properties[$key]->isModified())) {
-                                    $this->properties[$key]->_setLength($value);
-                                    if ($this->properties[$key]->isDirty()) {
-                                        $this->properties[$key]->setClean();
+                                $property = $this->loadProperty($key);
+                                if (! ($keepChanges && $property->isModified())) {
+                                    $property[$key]->_setLength($value);
```
Comment (dbu) [inline src/Jackalope/Node.php]: i think `[$key]` are wrong here (and the 2 lines below)

File: src/Jackalope/Node.php [hunk 2]
```diff
@@ -302,7 +310,9 @@ private function parseData($rawData, $update, $keepChanges = false)
                 $this->nodes = $nodesInBackend;
                 $this->originalNodesOrder = null;
             }
-            foreach ($oldProperties as $name => $property) {
+            foreach ($oldPropertyValues as $name => $propertyValues) {
+                // TODO needs an update
+                $property = $this->loadProperty($name);
```
Comment (dbu) [inline src/Jackalope/Node.php]: i think you try to load the new property here, where you should load the old data.

File: src/Jackalope/Node.php [hunk 3]
```diff
@@ -773,9 +785,10 @@ public function getPropertiesValues($nameFilter = null, $dereference = true)
         $this->checkState();
 
         // OPTIMIZE: do not create properties in constructor, go over array here
-        $names = self::filterNames($nameFilter, array_keys($this->properties));
+        $names = self::filterNames($nameFilter, array_keys($this->propertyValues));
         $result = array();
         foreach ($names as $name) {
+            $this->loadProperty($name);
```
Comment (dbu) [inline src/Jackalope/Node.php]: i think the idea of getPropertiesValues was that you could get the values without needing to instantiate the properties. do we still need this call to do cleanup?

File: src/Jackalope/Node.php [hunk 4]
```diff
@@ -1452,7 +1467,8 @@ public function unsetProperty($name)
      */
     public function confirmSaved()
     {
-        foreach ($this->properties as $property) {
+        foreach ($this->propertyValues as $name => $propertyValue) {
+            $property = $this->loadProperty($name);
             if ($property->isModified() || $property->isNew()) {
```
Comment (dbu) [inline src/Jackalope/Node.php]: i think you don't need loadProperty here. if it was just loaded, it will be neither new nor modified. you could loop over the already existing property objects for this.

File: src/Jackalope/Node.php [hunk 5]
```diff
@@ -1462,17 +1478,6 @@ public function confirmSaved()
     }
 
     /**
-     * In addition to calling parent method, tell all properties
-     */
-    public function setPath($path, $move = false)
```
Comment (dbu) [inline src/Jackalope/Node.php]: i think you still need this for the properties that are already instantiated.

File: src/Jackalope/Node.php [hunk 6]
```diff
@@ -1564,12 +1575,24 @@ public function getIterator()
...
      */
-    protected function _setProperty($name, $value, $type, $internal)
+    protected function loadProperty($name)
     {
+        if (isset($this->properties[$name])) {
+            return $this->properties[$name];
+        }
+
+        if (!isset($this->propertyValues[$name])) {
+            return;
```
Comment (dbu) [inline src/Jackalope/Node.php]: is this the expected behaviour? this makes the method return a mix of property and null. could we throw the not existing exception here?

File: src/Jackalope/Node.php [hunk 7]
```diff
@@ -1564,12 +1575,24 @@ public function getIterator()
...
+
         if ($name == '' | false !== strpos($name, '/')) {
             throw new InvalidArgumentException("The name '$name' is no valid property name");
         }
 
+        $value = $this->propertyValues[$name]['value'];
+        $type = $this->propertyValues[$name]['type'];
+        $internal = $this->propertyValues[$name]['internal'];
+
         if (!isset($this->properties[$name])) {
```
Comment (dbu) [inline src/Jackalope/Node.php]: you already have this check (or rather, the opposite) above. if we get here the property will not exist yet.

File: src/Jackalope/Node.php [hunk 8]
```diff
@@ -1608,7 +1633,8 @@ protected function _setProperty($name, $value, $type, $internal)
     public function setDirty($keepChanges = false, $targetState = false)
     {
         parent::setDirty($keepChanges, $targetState);
-        foreach ($this->properties as $property) {
+        foreach ($this->propertyValues as $name => $propertyValue) {
+            $property = $this->loadProperty($name);
```
Comment (dbu) [inline src/Jackalope/Node.php]: we only need to do this for already loaded properties. the ones not created yet do not need to be re-created later.

File: src/Jackalope/Node.php [hunk 9]
```diff
@@ -1639,7 +1665,8 @@ public function setChildrenDirty()
     public function setDeleted()
     {
         parent::setDeleted();
-        foreach ($this->properties as $property) {
+        foreach ($this->propertyValues as $name => $propertyValue) {
+            $property = $this->loadProperty($name);
```
Comment (dbu) [inline src/Jackalope/Node.php]: same here, only set existing property objects to deleted. all other objects did not exist so there can be no references to them.

File: src/Jackalope/Node.php [hunk 10]
```diff
@@ -1656,8 +1683,9 @@ public function beginTransaction()
         parent::beginTransaction();
 
         // Notify the children properties
-        foreach ($this->properties as $prop) {
-            $prop->beginTransaction();
+        foreach ($this->propertyValues as $name => $propertyValue) {
+            $property = $this->loadProperty($name);
```
Comment (dbu) [inline src/Jackalope/Node.php]: and again here, non existing properties don't need to know about this, i think. or do they?

**Comment by danrot** [comment]
@dbu I'll have a look at your comments soon, but atm I am struggling with the tests... I am getting 800 errors like this (even when running the tests on master):

```
DOMDocument::loadXML(): attributes construct error in Entity, line: 1
```

Any idea what can cause this?

**Comment by dbu** [comment]
dbal or jackrabbit? maybe the version of libxml?

**Comment by danrot** [comment]
dbal, I've got libxml 2.9.2 (according to the `php -i`), shouldn't that be fine?

File: src/Jackalope/Node.php (hunk 2 above)
Comment (danrot) [inline src/Jackalope/Node.php]: TBH, I don't really know what's happening here. But looks like you are right, so I can't use the `loadProperties` method here... What would be the easiest way to achieve that? Would it be enough to only use the nodes that have already been instantiated before the call of `parseData`?

File: src/Jackalope/Node.php (hunk 3 above)
Comment (danrot) [inline src/Jackalope/Node.php]: Since afterwards `$this->properties[$name]` is accessed, I was assuming that the property was already instantiated. And since before this PR all properties have been instantiated eagerly, this could have not been the purpose of this method (or at least it didn't work that way), or am I missing something?

File: src/Jackalope/Node.php (hunk 4 above)
Comment (danrot) [inline src/Jackalope/Node.php]: Ok, so I just revert these changes?

File: src/Jackalope/Node.php (hunk 5 above)
Comment (danrot) [inline src/Jackalope/Node.php]: I've understood my implementation in a way that `loadProperty` is called for every property (since all the `_setProperty` calls do not instantiate properties anymore). And I've moved the `setPath` call there, is there a scenario where this would not work?

File: src/Jackalope/Node.php (hunk 6 above)
Comment (danrot) [inline src/Jackalope/Node.php]: Had to put this here in order to be able to save nodes again... So I guess just throwing another exception would also not work... I'll try to find another solution.

File: src/Jackalope/Node.php (hunk 7 above)
Comment (danrot) [inline src/Jackalope/Node.php]: So that means I can delete the else path? I am not totally sure about this.

File: src/Jackalope/Node.php (hunk 10 above)
Comment (danrot) [inline src/Jackalope/Node.php]: I don't really know... Shouldn't a property that is added later have the same transaction state? However, if this is removed, it should also be removed in the `commitTransaction` and `rollbackTransaction` methods, right?

File: src/Jackalope/Node.php (hunk 3 above)
Comment (dbu) [inline src/Jackalope/Node.php]: sorry i was unclear. we added this method (among other reasons) to provide ourselves a way to *in the future* provide access to the values without needing to instantiate properties.

however, its probably a bit tricky, depending on the type (reference, date, ...) we would need to convert a string value so it could get really quite hairy. we could check if we have the property and if not, create it unless its a string/int/float property in which case we can just use the value as is, wdyt?

File: src/Jackalope/Node.php (hunk 4 above)
Comment (dbu) [inline src/Jackalope/Node.php]: yes please. maybe add a comment to avoid confusion in the future.

File: src/Jackalope/Node.php (hunk 5 above)
Comment (dbu) [inline src/Jackalope/Node.php]: at least when we have a move operation, we must propagate that to the already instantiated properties. i think you should just do the same we have now (no need to load not yet loaded properties)

File: src/Jackalope/Node.php (hunk 7 above)
Comment (dbu) [inline src/Jackalope/Node.php]: hm, maybe the github diff is confusing me here. but i would not see that you can ever reach the else path below. look at line 1580 - if the property is already set, you return it.

File: src/Jackalope/Node.php (hunk 10 above)
Comment (dbu) [inline src/Jackalope/Node.php]: yes. i guess what you however do need is if you load a property while inside a transaction, pass the transaction status to that property. otherwise the commit/rollback will be a problem on a property that was loaded while inside the transaction.

File: src/Jackalope/Node.php (hunk 2 above)
Comment (dbu) [inline src/Jackalope/Node.php]: this is part of the refresh operation. i think it was about removing newly created properties again if we refresh and they where not saved. so you could just have an $oldProperties with the already loaded properties and loop over those here. if they are newly created, they certainly are loaded.

File: src/Jackalope/Node.php [hunk 11]
```diff
@@ -213,9 +221,9 @@ private function parseData($rawData, $update, $keepChanges = false)
                     continue;
                 }
 
-                if ($update && array_key_exists($key, $this->properties)) {
-                    unset($oldProperties[$key]);
-                    $prop = $this->properties[$key];
+                if ($update && array_key_exists($key, $this->propertyValues)) {
+                    unset($oldPropertyValues[$key]);
+                    $prop = $this->loadProperty($key);
```
Comment (dbu) [inline src/Jackalope/Node.php]: i think same here, this is only about properties that are already loaded. if they are not loaded, they can't be modified or deleted (or does deleting a property using Node::setValue not lead to the property being loaded?)

**Comment by dbu** [comment]
it should. did the problem occur before your changes as well?

File: src/Jackalope/Node.php (hunk 3 above)
Comment (danrot) [inline src/Jackalope/Node.php]: How cheap is it to get the type of a property? At this place there's only the name, so I have to load something either way. I am not sure if this causes a notable performance gain, and therefore is worth the duplication of this logic.

File: src/Jackalope/Node.php (hunk 5 above)
Comment (danrot) [inline src/Jackalope/Node.php]: Yeah, that's what I thought, so I've moved this logic to https://github.com/jackalope/jackalope/pull/304/files#diff-194b8b91be1497a30eb4ca5256bc194bR1623

File: src/Jackalope/Node.php (hunk 7 above)
Comment (danrot) [inline src/Jackalope/Node.php]: I know, but I am sure that the code in the else path was there for a reason. Maybe there's a caveat in this new strategy? But if you say it should be save to delete I'll do it.

File: src/Jackalope/Node.php (hunk 10 above)
Comment (danrot) [inline src/Jackalope/Node.php]: Would you also do that in this PR? Or can that be a separate one? (since this one is more about read performance)

File: src/Jackalope/Node.php (hunk 11 above)
Comment (danrot) [inline src/Jackalope/Node.php]: Makes sense, will change it

**Comment by danrot** [comment]
Yeah, that's the confusing part... However, I don't get the error with jackrabbit, will test there first.

**Comment by danrot** [comment]
I think before you further review it makes sense that I get this completely running first. Might have to change some stuff anyway.

File: src/Jackalope/Node.php (hunk 5 above)
Comment (dbu) [inline src/Jackalope/Node.php]: assume this:

```php
$n = $session->getNode('/bar');
$p = $n->getProperty('foo');
$session->move('/bar', '/other');
echo $p->getPath();
```

we do still need setPath, but only update the properties that are already loaded

File: src/Jackalope/Node.php (hunk 7 above)
Comment (dbu) [inline src/Jackalope/Node.php]: the previous code was about _setProperty. that is not the same as loadProperty. not sure if we can eliminate _setProperty totally.

File: src/Jackalope/Node.php (hunk 10 above)
Comment (dbu) [inline src/Jackalope/Node.php]: i don't think there is much missing. just revert the changes on the transaction methods and pass the transaction status in loadProperty should do the trick. the problem is that it will become hard to see what is going on once refactored this way.

File: src/Jackalope/Node.php [hunk 12]
```diff
@@ -580,14 +588,23 @@ public function setProperty($name, $value, $type = PropertyType::UNDEFINED, $val
...
 
-        return $this->_setProperty($name, $value, $type, false);
+        $this->_setProperty($name, $value, $type, false);
+
+        if (isset($this->properties[$name])) {
+            $property = $this->properties[$name];
+            $property->setValue($value, $type ?: PropertyType::UNDEFINED);
+        } else {
+            $property = $this->loadProperty($name);
+        }
```
Comment (danrot) [inline src/Jackalope/Node.php]: I think I have to find a cleaner solution here... But for now I just try to get the tests with jackalope-jackrabbit running again, and clean up afterwards.

**Comment by danrot** [comment]
@dbu I am having troubles with the `testAddSessionRefresh` test in `CombinedManipulationsTest`. The problem is that I don't know how the property disappears on the call https://github.com/phpcr/phpcr-api-tests/blob/master/tests/Writing/CombinedManipulationsTest.php#L695. I've debugged it, and I realize that the `prop` property is just gone after executing this line: https://github.com/jackalope/jackalope/blob/master/src/Jackalope/Item.php#L613

It looks like a simple comparison, but yet the property seems to be deleted afterwards. How does this work?

**Comment by danrot** [comment]
ping @dbu

**Comment by akira28** [comment]
any news on this? We are hitting a wall importing several thousands pages, and could solve the performance issue we are facing.
@danrot Is your fork usable?

**Comment by akira28** [comment]
full disclosure, we are working on a Sulu project, so you should have a lot of insight on this particular issue :D

**Comment by dbu** [comment]
@danrot are you having a look into this? please rebase the branch. if you are still stuck on the failure, i can have another look at it next week when i am back from canada.

**Comment by akira28** [comment]
thanks @danrot for rebasing 👍

**Comment by danrot** [comment]
No problem. Have seen that the tests are already passing here, but there are still some problems when running the tests in jackalope-jackrabbit. Haven't tested jackalope-doctrine-dbal.

**Comment by akira28** [comment]
@dbu have you had the opportunity to look at it now that's rebased?

**Comment by dbu** [comment]
@danrot can you create a pull request on jackalope-jackrabbit and jackalope-doctrine-dbal that use this branch instead of the normal source (you will need to custom set where the repository is found) so that we can see on travisci how things look?

**Comment by danrot** [comment]
@dbu Created https://github.com/jackalope/jackalope-jackrabbit/pull/140 and https://github.com/jackalope/jackalope-doctrine-dbal/pull/331

**Comment by danrot** [comment]
Actually it doesn't look tooo bad 😃

**Comment by dbu** [comment]
yeah, seems to be mainly a couple of things with versioning and the combined manipulations.

can you go over the code again with a fresh mind and check in each place whether we should check the instantiated properties or the raw values, and make sure we always make sure to instantiate the property if needed? i feel there are some mixups what we use when. that could also explain why values are not updated in some situations.

also, diff calculation in transport should have a way to only look at properties that actually changed instead of creating instantiation of all of them. but that can be a separate pull request that comes later - its also kindof a BC break so we would need to check from transport if the jackalope version is new enough.

**Comment by danrot** [comment]
Sadly I am busy with other stuff currently, can't promise a due date to that 😕

**Comment by dbu** [comment]
ping @danrot

**Comment by danrot** [comment]
Sry, I really don't have any time available for working at this at the moment 😕
